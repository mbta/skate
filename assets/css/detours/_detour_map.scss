.c-detour_map--original-route-shape {
  stroke: $color-kiwi-500;
}

.c-detour_map--original-route-shape__unstarted {
  cursor: url("data:image/svg+xml,%3Csvg width='22' height='22' viewBox='0 0 22 22' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cstyle%3E .color %7B fill: %2303A900; %7D %3C/style%3E%3Ccircle cx='11' cy='11' r='10' class='color' style='filter: opacity(0.5) blur(0.5px)'/%3E%3Ccircle cx='11' cy='11' r='7' class='color' stroke='white' stroke-width='2'/%3E%3C/svg%3E")
      11 11,
    pointer;
}

.c-detour_map--detour-route-shape {
  stroke: $color-lemon-500;
}

.c-detour_map-circle-marker--start {
  stroke: $color-kiwi-400;
  fill: $color-kiwi-400;
}

.c-detour_map-circle-marker--end {
  stroke: $color-strawberry-400;
  fill: $color-strawberry-400;
}
