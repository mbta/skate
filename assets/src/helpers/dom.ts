// ts-lint doesn't like either Array<string | undefined> or
// (string | undefined)[], so we create this type as a workaround.
export type MaybeString = string | undefined

export const className = (classes: MaybeString[]): string =>
  classes.filter((c) => c && c !== "").join(" ")
