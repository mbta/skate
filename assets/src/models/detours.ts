export interface DraftDetour {
  route: string
  direction: string
  name: string
  intersection: string
  activeSince: number
}

export type DraftDetours = DraftDetour[]
