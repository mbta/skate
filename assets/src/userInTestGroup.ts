import getTestGroups from "./userTestGroups"

export enum TestGroups {
  DemoMode = "demo-mode",
  DetoursPilot = "detours-pilot",
  DummyDetourPage = "dummy-detour-page",
  KeycloakSso = "keycloak-sso",
  MinimalLadderPage = "minimal-ladder-page",
  LateView = "late-view",
  RouteLadderHeaderUpdate = "route-ladder-header-update",
}

const inTestGroup = (key: TestGroups): boolean => {
  return getTestGroups().includes(key)
}

export default inTestGroup
