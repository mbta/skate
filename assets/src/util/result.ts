/**
 * The discriminator value used to determine if a {@linkcode Result<T, E>} is
 * {@linkcode Ok<T>} or {@linkcode Err<E>}.
 */
enum ResultTag {
  Ok,
  Err,
}

/**
 * The key of the {@linkcode Result<T, E>} discriminator value.
 *
 * This is not exported so that usages of the objects cannot be
 * accessed or modified manually outside of this module.
 */
const tag = Symbol("result-enum-tag")

/** The {@linkcode Ok<T>} variant of {@linkcode Result<T, E>} */
export type Ok<T> = { [tag]: ResultTag.Ok; ok: T }

/** The {@linkcode Err<E>} variant of {@linkcode Result<T, E>} */
export type Err<E> = { [tag]: ResultTag.Err; err: E }

/**
 * Type which may contain an {@linkcode Ok<T>} state, or a {@linkcode Err<E>} state
 *
 * useful for requiring a caller of a function to handle the Error state of a return value.
 */
export type Result<T, E> = Ok<T> | Err<E>

/** Determines if a {@linkcode Result<T, E>} is a {@linkcode Ok<T>} */
export const isOk = <T, E>(r: Result<T, E>): r is Ok<T> =>
  tag in r && r[tag] === ResultTag.Ok

/** Determines if a {@linkcode Result<T, E>} is a {@linkcode Err<E>} */
export const isErr = <T, E>(r: Result<T, E>): r is Err<E> =>
  tag in r && r[tag] === ResultTag.Err

/** Constructs a {@linkcode Ok<T>} with {@linkcode v} as the value */
export const Ok = <T>(v: T): Ok<T> => ({
  [tag]: ResultTag.Ok,
  ok: v,
})

/** Constructs a {@linkcode Err<E>} with {@linkcode v} as the value */
export const Err = <E>(v: E): Err<E> => ({
  [tag]: ResultTag.Err,
  err: v,
})

/**
 * If {@linkcode r} is {@linkcode Ok<T>}, it applies {@linkcode fn} to
 * the value contained in {@linkcode r}. Otherwise returns {@linkcode r}.
 */
export const map = <T, U, E>(r: Result<T, E>, fn: (v: T) => U): Result<U, E> =>
  isOk(r) ? Ok(fn(r.ok)) : r
