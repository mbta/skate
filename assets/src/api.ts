import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
