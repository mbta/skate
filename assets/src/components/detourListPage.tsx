import React, { useCallback, useState } from "react"
import { DetoursTable, DetourStatus } from "./detoursTable"
import userInTestGroup, { TestGroups } from "../userInTestGroup"
import { Button } from "react-bootstrap"
import { PlusSquare } from "../helpers/bsIcons"
import { DetourModal } from "./detours/detourModal"
import { fetchDetour, fetchDetours } from "../api"
import { useApiCall } from "../hooks/useApiCall"
import { isErr, isOk } from "../util/result"
import { isValidSnapshot } from "../util/isValidSnapshot"
import { createDetourMachine } from "../models/createDetourMachine"

export const DetourListPage = () => {
  const [showDetourModal, setShowDetourModal] = useState(false)
  const [detourId, setDetourId] = useState<number | undefined>()
  const { result } = useApiCall({
    apiCall: useCallback(async () => fetchDetours(), []),
  })
  const detours = result && isOk(result) && result.ok

  const { result: stateOfDetourModal } = useApiCall({
    apiCall: useCallback(async () => {
      if (detourId === undefined) {
        return undefined
      }
      const detourResponse = await fetchDetour(detourId)
      if (isErr(detourResponse)) {
        return undefined
      }
      const snapshot = isValidSnapshot(
        createDetourMachine,
        detourResponse.ok.state
      )
      if (isErr(snapshot)) {
        return undefined
      }
      return snapshot.ok
    }, [detourId]),
  })

  const onOpenDetour = (detourId: number) => {
    setDetourId(detourId)
    setShowDetourModal(true)
  }

  const onCloseDetour = () => {
    setDetourId(undefined)
    setShowDetourModal(false)
  }

  return (
    <div className="c-detour-list-page h-100 overflow-y-auto p-0 p-md-4 bg-white">
      {userInTestGroup(TestGroups.DetoursPilot) && (
        <Button
          className="c-detour-list-page__button icon-link fw-light px-3 py-2 u-hide-for-mobile"
          onClick={() => setShowDetourModal(true)}
        >
          <PlusSquare />
          <span className="c-detour-list-page__button-text">Add detour</span>
        </Button>
      )}
      {detours && (
        <>
          <Title title="Active detours" />
          <DetoursTable
            data={detours.active}
            status={DetourStatus.Active}
            onOpenDetour={onOpenDetour}
            classNames={["mb-5"]}
          />
          <Title title="Draft detours" />
          <DetoursTable
            data={detours.draft}
            status={DetourStatus.Draft}
            onOpenDetour={onOpenDetour}
            classNames={["mb-5"]}
          />
          <Title title="Closed detours" />
          <DetoursTable
            data={detours.past}
            status={DetourStatus.Closed}
            onOpenDetour={onOpenDetour}
          />
        </>
      )}

      {/* `detourId` exists before `stateOfDetourModal` does, so need this conditional
       * to ensure that either there's no `detourId` selected (i.e. make a new detour)
       * or the state has been successfully fetched from the api
       */}
      {showDetourModal && (!detourId || stateOfDetourModal) && (
        <DetourModal
          onClose={onCloseDetour}
          show
          originalRoute={{}}
          key={detourId ?? ""}
          {...(stateOfDetourModal ? { snapshot: stateOfDetourModal } : {})}
        />
      )}
    </div>
  )
}

const Title = ({ title }: { title: string }) => (
  <h2 className="fw-semibold fs-1 mt-3 mt-md-0 mb-3 mx-3 mx-md-0">{title}</h2>
)
