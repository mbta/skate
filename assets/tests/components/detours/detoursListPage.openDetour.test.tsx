import { beforeEach, describe, expect, jest, test } from "@jest/globals"

import React from "react"

import { screen, waitFor } from "@testing-library/dom"
import "@testing-library/jest-dom/jest-globals"
import { render } from "@testing-library/react"
import userEvent from "@testing-library/user-event"

import { DetourListPage } from "../../../src/components/detourListPage"
import { fetchDetour, fetchDetours } from "../../../src/api"
import { Ok } from "../../../src/util/result"
import { mockScreenSize, neverPromise } from "../../testHelpers/mockHelpers"
import getTestGroups from "../../../src/userTestGroups"
import { detourListFactory } from "../../factories/detourListFactory"
import { TestGroups } from "../../../src/userInTestGroup"
import { detourStateMachineFactory } from "../../factories/detourStateMachineFactory"
import { viewDraftDetourHeading } from "../../testHelpers/selectors/components/detours/diversionPage"

jest
  .useFakeTimers({ doNotFake: ["setTimeout"] })
  .setSystemTime(new Date("2024-08-29T20:00:00"))

jest.mock("../../../src/userTestGroups")

jest.mock("../../../src/api")

beforeEach(() => {
  jest.mocked(fetchDetours).mockReturnValue(neverPromise())
  jest.mocked(fetchDetour).mockReturnValue(neverPromise())
  jest.mocked(getTestGroups).mockReturnValue([TestGroups.DetoursPilot])
})

describe("Detours Page: Open a Detour", () => {
  test("calls API with correct detour ID", async () => {
    jest.mocked(fetchDetours).mockResolvedValue(Ok(detourListFactory.build()))

    render(<DetourListPage />)

    await userEvent.click(await screen.findByText("Headsign 1"))
    expect(fetchDetour).toHaveBeenCalledWith(1)
  })

  test("renders detour details modal to match mocked fetchDetour", async () => {
    jest.mocked(fetchDetours).mockResolvedValue(Ok(detourListFactory.build()))

    // Return the state of the machine as the fetchDetour mocked value,
    // even if it doesn't match the detour clicked
    jest
      .mocked(fetchDetour)
      .mockResolvedValue(Ok(detourStateMachineFactory.build()))

    const { baseElement } = render(<DetourListPage />)

    // Click an arbitrary detour from the list
    await userEvent.click(await screen.findByText("Headsign Z"))

    // Render modal based on mocked value, which is a detour-in-progress
    expect(viewDraftDetourHeading.get()).toBeVisible()

    // Finally, check snapshot
    await waitFor(() => expect(baseElement).toMatchSnapshot())
  })

  test("renders detour details in an open drawer on mobile", async () => {
    mockScreenSize("mobile")

    jest.mocked(fetchDetours).mockResolvedValue(Ok(detourListFactory.build()))

    jest
      .mocked(fetchDetour)
      .mockResolvedValue(Ok(detourStateMachineFactory.build()))

    const { baseElement } = render(<DetourListPage />)

    await userEvent.click(await screen.findByText("Headsign Z"))

    await waitFor(() => expect(baseElement).toMatchSnapshot())
  })

  test("detour details drawer is collapsible on mobile", async () => {
    mockScreenSize("mobile")

    jest.mocked(fetchDetours).mockResolvedValue(Ok(detourListFactory.build()))

    jest
      .mocked(fetchDetour)
      .mockResolvedValue(Ok(detourStateMachineFactory.build()))

    render(<DetourListPage />)

    await userEvent.click(await screen.findByText("Headsign Z"))
    await userEvent.click(await screen.findByTitle("Collapse"))
    expect(await screen.findByTitle("Expand")).toBeVisible()
  })
})
