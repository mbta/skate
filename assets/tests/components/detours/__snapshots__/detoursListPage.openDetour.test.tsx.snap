// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detours Page: Open a Detour renders detour details modal to match mocked fetchDetour 1`] = `
<body
  data-rr-ui-modal-open=""
  style="padding-right: 1024px;"
>
  <div>
    <div
      class="c-detour-list-page h-100 overflow-y-auto p-0 p-md-4 bg-white"
    >
      <table
        class="c-detours-table table table-hover"
      >
        <thead
          class="u-hide-for-mobile"
        >
          <tr>
            <th
              class="px-3 py-4"
            >
              Route and direction
            </th>
            <th
              class="px-3 py-4 u-hide-for-mobile"
            >
              Starting Intersection
            </th>
            <th
              class="px-3 py-4 u-hide-for-mobile"
            >
              On detour since
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="align-middle p-3"
            >
              <div
                class="d-flex"
              >
                <div
                  class="c-route-pill c-route-pill--bus"
                >
                  4
                </div>
                <div
                  class="c-detours-table__route-info-text d-inline-block"
                >
                  <div
                    class="pb-1 fs-4 fw-semibold"
                  >
                    Headsign 4
                  </div>
                  <div
                    class="c-detours-table__route-info-direction fs-6"
                  >
                    Inbound
                  </div>
                </div>
              </div>
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              Street A4 & Avenue B4
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              26 hours ago
            </td>
          </tr>
          <tr>
            <td
              class="align-middle p-3"
            >
              <div
                class="d-flex"
              >
                <div
                  class="c-route-pill c-route-pill--bus"
                >
                  5
                </div>
                <div
                  class="c-detours-table__route-info-text d-inline-block"
                >
                  <div
                    class="pb-1 fs-4 fw-semibold"
                  >
                    Headsign 5
                  </div>
                  <div
                    class="c-detours-table__route-info-direction fs-6"
                  >
                    Outbound
                  </div>
                </div>
              </div>
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              Street A5 & Avenue B5
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              26 hours ago
            </td>
          </tr>
        </tbody>
      </table>
      <table
        class="c-detours-table table table-hover"
      >
        <thead
          class="u-hide-for-mobile"
        >
          <tr>
            <th
              class="px-3 py-4"
            >
              Route and direction
            </th>
            <th
              class="px-3 py-4 u-hide-for-mobile"
            >
              Starting Intersection
            </th>
            <th
              class="px-3 py-4 u-hide-for-mobile"
            >
              On detour since
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="align-middle p-3"
            >
              <div
                class="d-flex"
              >
                <div
                  class="c-route-pill c-route-pill--bus"
                >
                  6
                </div>
                <div
                  class="c-detours-table__route-info-text d-inline-block"
                >
                  <div
                    class="pb-1 fs-4 fw-semibold"
                  >
                    Headsign Z
                  </div>
                  <div
                    class="c-detours-table__route-info-direction fs-6"
                  >
                    Inbound
                  </div>
                </div>
              </div>
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              Street A6 & Avenue B6
            </td>
            <td
              class="align-middle p-3 u-hide-for-mobile"
            >
              26 hours ago
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div />
  <div
    aria-modal="true"
    class="c-detour-modal c-detour-modal__transition-appear c-detour-modal__transition-appear-active"
    role="dialog"
    tabindex="-1"
  >
    <article
      class="l-diversion-page h-100 border-box inherit-box"
    >
      <header
        class="l-diversion-page__header border-bottom text-bg-light"
      >
        <button
          aria-label="Close"
          class="btn-close p-4"
          type="button"
        />
      </header>
      <div
        class="l-diversion-page__panel text-bg-light"
      >
        <article
          class="l-diversion-page-panel border-end"
        >
          <header
            class="l-diversion-page-panel__header border-bottom px-3"
          >
            <h1
              class="c-diversion-panel__h1 my-3"
            >
              Share Detour Details
            </h1>
          </header>
          <div
            class="l-diversion-page-panel__body bg-light d-flex flex-column"
          >
            <div
              class="l-diversion-page-panel__scroll-area px-3 d-flex flex-column"
            >
              <button
                class="align-self-start icon-link my-3 btn btn-outline-primary"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="bi bi-arrow-left"
                  fill="currentColor"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M15 8a.5.5 0 0 0-.5-.5H2.707l3.147-3.146a.5.5 0 1 0-.708-.708l-4 4a.5.5 0 0 0 0 .708l4 4a.5.5 0 0 0 .708-.708L2.707 8.5H14.5A.5.5 0 0 0 15 8"
                    fill-rule="evenodd"
                  />
                </svg>
                 Edit Detour
              </button>
              <textarea
                class="flex-grow-1 mb-3 form-control"
                style="resize: none;"
              />
            </div>
            <div
              class="border-top d-flex mt-auto d-flex flex-column"
            >
              <button
                class="m-3 flex-grow-1 btn btn-primary"
                type="button"
              >
                Copy Details
              </button>
            </div>
          </div>
        </article>
      </div>
      <div
        class="l-diversion-page__map position-relative"
      >
        <div
          class="fade position-absolute top-0 left-0 m-2 icon-link z-1 alert alert-info show"
          role="alert"
        >
          <svg
            aria-hidden="true"
            class="bi bi-exclamation-circle-fill"
            fill="currentColor"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M16 8A8 8 0 1 1 0 8a8 8 0 0 1 16 0M8 4a.905.905 0 0 0-.9.995l.35 3.507a.552.552 0 0 0 1.1 0l.35-3.507A.905.905 0 0 0 8 4m.002 6a1 1 0 1 0 0 2 1 1 0 0 0 0-2"
            />
          </svg>
          Detour shape is not editable from this screen.
        </div>
        <div
          class="h-100"
        >
          <div
            class="c-vehicle-map-state"
          />
          <div
            class="c-vehicle-map leaflet-container leaflet-touch leaflet-grab leaflet-touch-drag leaflet-touch-zoom"
            id="id-vehicle-map"
            style="position: relative;"
            tabindex="0"
          >
            <div
              class="leaflet-pane leaflet-map-pane"
              style="left: 0px; top: 0px;"
            >
              <div
                class="leaflet-pane leaflet-tile-pane"
              >
                <div
                  class="leaflet-layer "
                  style="z-index: 1;"
                >
                  <div
                    class="leaflet-tile-container leaflet-zoom-animated"
                    style="z-index: 21; left: 0px; top: 0px;"
                  >
                    <img
                      alt=""
                      class="leaflet-tile"
                      src="undefined"
                      style="width: 256px; height: 256px; left: -4px; top: -239px;"
                    />
                  </div>
                </div>
              </div>
              <div
                class="leaflet-pane leaflet-overlay-pane"
              >
                <svg
                  height="0"
                  pointer-events="none"
                  style="left: 0px; top: 0px;"
                  viewBox="0 0 0 0"
                  width="0"
                >
                  <g>
                    <path
                      class="c-detour_map--detour-route-shape"
                      d="M0 0"
                      fill="none"
                      stroke="#3388ff"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-opacity="1"
                      stroke-width="6"
                    />
                    <path
                      class="c-detour_map--original-route-shape-core"
                      d="M0 0"
                      fill="none"
                      stroke="#3388ff"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-opacity="1"
                      stroke-width="6"
                    />
                  </g>
                </svg>
              </div>
              <div
                class="leaflet-pane leaflet-shadow-pane"
              />
              <div
                class="leaflet-pane leaflet-marker-pane"
              >
                <div
                  class="leaflet-pane leaflet-primaryVehicles-pane"
                  style="z-index: 499;"
                />
                <div
                  class="leaflet-pane leaflet-secondaryVehicles-pane"
                  style="z-index: 400;"
                />
                <div
                  class="leaflet-pane leaflet-notableLocationMarkers-pane"
                  style="z-index: 410;"
                />
                <div
                  class="leaflet-marker-icon leaflet-zoom-hide"
                  role="button"
                  style="margin-left: -8px; margin-top: -8px; width: 16px; height: 16px; left: 0px; top: 0px; z-index: 0;"
                  tabindex="0"
                  title="Detour Start"
                >
                  <div
                    class="w-100 h-100"
                  >
                    <svg
                      class="c-detour_map-circle-marker--start"
                      height="100%"
                      viewBox="0 0 16 16"
                      width="100%"
                    >
                      <circle
                        cx="8"
                        cy="8"
                        opacity="0.5"
                        r="7.5"
                      />
                      <circle
                        cx="8"
                        cy="8"
                        r="6"
                        stroke="white"
                        stroke-width="2"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="leaflet-marker-icon leaflet-zoom-hide"
                  role="button"
                  style="margin-left: -5px; margin-top: -5px; width: 10px; height: 10px; left: 0px; top: 0px; z-index: 0;"
                  tabindex="0"
                >
                  <div
                    class="w-100 h-100"
                  >
                    <svg
                      class="c-detour_map-circle-marker--detour-point"
                      fill="none"
                      height="100%"
                      viewBox="0 0 10 10"
                      width="100%"
                    >
                      <circle
                        cx="5"
                        cy="5"
                        r="4.5"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="leaflet-marker-icon leaflet-zoom-hide"
                  role="button"
                  style="margin-left: -8px; margin-top: -8px; width: 16px; height: 16px; left: 0px; top: -1px; z-index: -1;"
                  tabindex="0"
                  title="Detour End"
                >
                  <div
                    class="w-100 h-100"
                  >
                    <svg
                      class="c-detour_map-circle-marker--end"
                      height="100%"
                      viewBox="0 0 16 16"
                      width="100%"
                    >
                      <circle
                        cx="8"
                        cy="8"
                        opacity="0.5"
                        r="7.5"
                      />
                      <circle
                        cx="8"
                        cy="8"
                        r="6"
                        stroke="white"
                        stroke-width="2"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="leaflet-marker-icon c-stop-icon-container c-vehicle-map__stop leaflet-zoom-hide leaflet-interactive"
                  role="button"
                  style="margin-left: -7px; margin-top: -7px; width: 14px; height: 14px; left: 0px; top: 0px; z-index: 0;"
                  tabindex="0"
                >
                  <div
                    class="w-100 h-100"
                  >
                    <svg
                      class="c-stop-icon"
                      height="100%"
                      width="100%"
                    >
                      <svg
                        fill="none"
                        viewBox="0 0 14 14"
                      >
                        <circle
                          class="c-stop-icon__focus-circle"
                          cx="7"
                          cy="7"
                          r="8"
                        />
                        <circle
                          class="c-stop-icon__circle c-stop-icon__circle--small"
                          cx="7"
                          cy="7"
                          r="5"
                        />
                      </svg>
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="leaflet-pane leaflet-tooltip-pane"
              />
              <div
                class="leaflet-pane leaflet-popup-pane"
              />
            </div>
            <div
              class="leaflet-control-container"
            >
              <div
                class="leaflet-top leaflet-left"
              />
              <div
                class="leaflet-top leaflet-right"
              >
                <div
                  class="leaflet-control-zoom leaflet-bar leaflet-control"
                >
                  <a
                    aria-disabled="false"
                    aria-label="Zoom in"
                    class="leaflet-control-zoom-in"
                    href="#"
                    role="button"
                    title="Zoom in"
                  >
                    <span
                      aria-hidden="true"
                    >
                      +
                    </span>
                  </a>
                  <a
                    aria-disabled="false"
                    aria-label="Zoom out"
                    class="leaflet-control-zoom-out"
                    href="#"
                    role="button"
                    title="Zoom out"
                  >
                    <span
                      aria-hidden="true"
                    >
                      −
                    </span>
                  </a>
                </div>
                <div
                  class="leaflet-bar leaflet-control"
                >
                  <a
                    aria-label="Full Screen"
                    class="leaflet-control-zoom-fullscreen fullscreen-icon"
                    href="#"
                    role="button"
                    title="Full Screen"
                  />
                </div>
              </div>
              <div
                class="leaflet-bottom leaflet-left"
              />
              <div
                class="leaflet-bottom leaflet-right"
              >
                <div
                  class="leaflet-control-attribution leaflet-control"
                >
                  © 
                  <a
                    href="http://osm.org/copyright"
                  >
                    OpenStreetMap
                  </a>
                   contributors
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </article>
  </div>
</body>
`;
